### DISCUSSION TITLE
TypeScript types for graphql function in createPages strange behaviour

### DISCUSSION THREAD

**Discussion by trevorblades** [gatsbyjs/gatsby#36144] [Help] [answered]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

The type of the `graphql` helper from `createPages` seems to change between where I destructure it to when it is invoked. Unfortunately, that makes the returned `data` property take on an `unknown` type, which makes it difficult to do anything with it without TypeScript complaining.

![f6e865392366aacb409eeaa75df2dbe1](https://user-images.githubusercontent.com/1216917/179305990-c3acf682-892b-4398-a28c-b38572b98645.gif)

I made a [very minimal reproduction of this issue](https://codesandbox.io/s/happy-firefly-vzpix3?file=/gatsby-node.ts) to test it yourself. Is there something more that I should be doing to have these query results typed properly?

Here's what the example `gatsby-node.ts` file looks like:

```ts
import { GatsbyNode } from "gatsby";

export const createPages: GatsbyNode["createPages"] = async ({
  graphql,
  actions
}) => {
  const { data } = await graphql(`
    query MyQuery {
      allSitePage {
        nodes {
          path
        }
      }
    }
  `);

  data.allSitePage.nodes.forEach((node) => {
    console.log(node.path);
  });
};
```

### Reproduction Link

https://codesandbox.io/s/happy-firefly-vzpix3?file=/gatsby-node.ts

### Steps to Reproduce

1. Export a `createPages` function from `gatsby-node.ts`.
2. Use the `graphql` helper function to query for some data.
3. Observe the difference in typing between the `graphql` described in `CreatePagesArgs` and the `graphql` function being invoked.

### Expected Result

I expect the signature of `graphql` to not change, and for `data` to be typed as `TData | undefined`, as it is [in its declaration](https://github.com/gatsbyjs/gatsby/blob/2293a758a867415e6ae757e6ba8c6db351548bf1/packages/gatsby/index.d.ts#L805-L811).

### Actual Result

`data` is typed as `unknown`, and TypeScript gives me a hard time about doing basically anything with it, including `data?.allCourse.nodes.forEach`.

### Environment

```shell
System:
    OS: macOS 11.6.7
    CPU: (8) x64 Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.15.0 - ~/.volta/tools/image/node/16.15.0/bin/node
    Yarn: 1.22.17 - ~/.volta/tools/image/yarn/1.22.17/bin/yarn
    npm: 8.10.0 - ~/.volta/tools/image/npm/8.10.0/bin/npm
  Languages:
    Python: 2.7.11 - /usr/local/bin/python
  Browsers:
    Chrome: 103.0.5060.114
    Firefox: 97.0.2
    Safari: 15.5
  npmPackages:
    gatsby: ^4.18.2 => 4.18.2 
    gatsby-plugin-apollo: ^4.0.2 => 4.0.2 
    gatsby-plugin-google-gtag: ^4.16.0 => 4.16.0 
    gatsby-plugin-lodash: ^5.16.0 => 5.16.0 
    gatsby-plugin-mdx: ^3.15.0 => 3.15.0 
    gatsby-plugin-next-seo: ^1.10.0 => 1.10.0 
    gatsby-plugin-svgr: ^3.0.0-beta.0 => 3.0.0-beta.0 
    gatsby-plugin-webfonts: ^2.2.2 => 2.2.2 
    gatsby-source-filesystem: ^4.15.0 => 4.15.0
```

**Comment by LekoArts** [comment, marked as answer]
TData is a generic: https://www.typescriptlang.org/docs/handbook/2/generics.html

If you don’t define it it'll become unknown. Pass in your type to the function and your result will be typed, e.g. like in https://github.com/LekoArts/portfolio-v2/blob/main/www/gatsby-node.ts#L33

```ts
const result = await graphql<YourType>()
```

**Reply by trevorblades** [reply]
Thank you! This is my first time using this pattern. I learned something new today 🙂
